README.md

```diff
@@ -11,572 +11,528 @@
+main :: IO ()
+main = withFile "output.txt" AppendMode server
 ```
 
-The concurrency facilities provided by streamly can be compared with
-[OpenMP](https://en.wikipedia.org/wiki/OpenMP) and
-[Cilk](https://en.wikipedia.org/wiki/Cilk) but with a more declarative
-expression.
+## Listing Directories Recursively/Concurrently
+
+The following example lists a directory tree recursively, reading
+multiple directories concurrently.
 
-## Example: Listing Directories Recursively/Concurrently
+It uses the wonderful tree traversing combinator
+`iterateMapLeftsWith`. It maps a stream generator on the `Left` values
+(directories in this case) of the input stream, feeds the resulting
+'Left' values back to the input, and lets the `Right` values (files in
+this case) pass through to the output. The `Stream.ahead` stream joining
+combinator makes it iterate on the directories concurrently.
 
-The following code snippet lists a directory tree recursively, reading multiple
-directories concurrently:
+See
+[ListDir.hs](https://github.com/composewell/streamly-examples/blob/master/examples/ListDir.hs)
+for full working code including the imports that we may have omitted below.
 
 ```haskell
-import Control.Monad.IO.Class (liftIO)
-import Path.IO (listDir, getCurrentDir) -- from path-io package
-import Streamly.Prelude (AsyncT, adapt)
-import qualified Streamly.Prelude as S
-
-listDirRecursive :: AsyncT IO ()
-listDirRecursive = getCurrentDir >>= readdir >>= liftIO . mapM_ putStrLn
-  where
-    readdir dir = do
-      (dirs, files) <- listDir dir
-      S.yield (map show dirs ++ map show files) <> foldMap readdir dirs
+...
+import Streamly.Internal.Data.Stream.IsStream (iterateMapLeftsWith)
 
+import qualified Streamly.Prelude as Stream
+import qualified Streamly.Internal.FileSystem.Dir as Dir (toEither)
+
+-- Lists a dir as a stream of (Either Dir File)
+listDir :: String -> SerialT IO (Either String String)
+listDir dir =
+      Dir.toEither dir               -- SerialT IO (Either String String)
+    & Stream.map (bimap mkAbs mkAbs) -- SerialT IO (Either String String)
+
+    where mkAbs x = dir ++ "/" ++ x
+
+-- | List the current directory recursively using concurrent processing
 main :: IO ()
-main = S.drain $ adapt $ listDirRecursive
+main = do
+    hSetBuffering stdout LineBuffering
+    let start = Stream.yield (Left ".")
+    Stream.iterateMapLeftsWith Stream.ahead listDir start
+        & Stream.mapM_ print
 ```
 
-`AsyncT` is a stream monad transformer. If you are familiar with a list
-transformer, it is nothing but `ListT` with concurrency semantics. For example,
-the semigroup operation `<>` is concurrent. This makes `foldMap` concurrent
-too. You can replace `AsyncT` with `SerialT` and the above code will become
-serial, exactly equivalent to a `ListT`.
+Notice, how succinctly and idiomatically we expressed the concurrent
+directory tree traversal problem. Let us know if you can do it
+significantly better with any other language or framework.
```

Comment:
This can be told commonly for all examples in the end. We can go with the assumption that folk generally don't read the text surrounding the example.

Reply:
ok.